### ISSUE TITLE
GML: Specify callback functions from GML

### ISSUE THREAD

**Issue by kleinesfilmroellchen** [SerenityOS/serenity#13240]
The purpose of this issue is to collect feedback on whether this is a desirable feature. TL;DR: For widget callbacks, such as on_click, specify a function to attach to it straight from GML via the C++ function name.

# Goals
- Make it possible to link GUI to logic from GML in some capacity.
- Make the user effort for this to work as small as possible.
- Modify GML syntax as little as possible.
- Soft failure; i.e. if the attached-to function doesn't exist, nothing is attached instead and a warning is printed.

# Non-Goals
- Allow attaching to arbitrary C++ code from GML.
- Include C++ code in GML.
- Execute arbitrary C++ during GML creation aside from already executed code in the instantiated objects.

# Concept
We'll extend the syntax for GML values ([GUI::GML::ValueNode](https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibGUI/GML/AST.h#L60)) to the new *FunctionReference* type:

```ebnf
FunctionReference = CppIdentifier "()"
```

`CppIdentifier` is something akin to a valid C++ identifier, but a symbolic name. The actual signature of the function must match the signature of the callback, and this is checked when attaching the function.

For a function to be accessible via a function reference, it must be registered with a `REGISTER_GML_FUNCTION(name, function)` macro that works similar to `REGISTER_WIDGET`. The name is the term that the function reference can use, converted to a string internally. There's a global function registry (probably somewhere in Widget.cpp) that is used for looking up the functions when they are referenced in GML. As with Widget classes, this lookup happens at GML load time. If the signature is correct and the function exists, we assign it to the property through the existing property system. The property system needs to support non-JSON values for this.

This will look something like this:

```gml
@GUI::Button {
    text: "Search..."
    on_click: on_search_button_click()
}
```

```cpp
main() {
    REGISTER_GML_FUNCTION(on_search_button_click, [&](auto event) {
        // do stuff
    });
    load_from_gml(...);
}
```

The intention is that applications can then define a list of callback functions that are all registered, which can be capturing lambdas as is commonly needed, and after loading GML the functions are automatically attached.

## Improvements over the status quo
- Allow moving around objects in GML without needing to adjust the function attachment code.
- Easily replace GUI-attached functions by registering a different callback (also convenient for testing).
- No loss of readability: C++ function name and symbolic GML name appear within one macro call, making it easy to look up how the function can be used in GML.
- GML and GUI functionality are easier to reason about: In apps with GML, it's easier to see which widget has which callbacks attached and (with good names) what those callbacks do.
- Removal of boilerplate to reach into widget structures to attach callbacks.

In summary, while there are no extra powers or conciseness gained, this should decrease the mental load for GML GUI.

A main point of criticism is that we lose compile-time verification of function signatures. I don't think this is a big problem, as GML already loses us verification of property types (see e.g. 7459bac32346b0883a0c97e4baa819a33057b9d0) but we haven't seen any drawbacks from this. After all, GML verification happens on startup time, which is the earliest we get at run-time.

# Revision notes
Rev. 2: @alimpfard suggested the symbolic name methodology + support for capturing lambdas, which is very crucial and which I forgot about.
